design/complex-workloads.md

```diff
@@ -0,0 +1,988 @@
+library upon which Crossplane is built [does not expose the old version of an
+object during updates][15], but [validating webhooks do][16].
+
+Crossplane does not currently leverage Kubernetes webhooks, controller-runtime
+has [support for both validating and mutating admission webhooks][17]. This
+document proposes two validating webhook be added to Crossplane; one each of
+`KubernetesApplication` and `KubernetesApplicationComponent` to enforce
+immutability of the aforementioned fields.
+
+## Alternatives Considered
+The following alternative designs were considered and discarded or deferred in
+favor of the design proposed by this document.
+
+### Monolithic Workloads
+This alternative proposes a 'monolithic' workload. A monolithic workload is
+similar to the design proposed by this document but with the various templates
+and statuses nested directly within the `KubernetesApplication` rather than
+associated with the it via label selectors.
+
+An example monolithic complex workload:
+```yaml
+---
+apiVersion: workload.crossplane.io/v1alpha1
+kind: KubernetesApplication
+metadata:
+  name: demo
+spec:
+  clusterSelector:
+    provider: gcp
+  resources:
+  - name: demo
+    secretName: demo
+  templates:
+  - apiVersion: extensions/v1beta1
+    kind: Deployment
+    metadata:
+      name: wordpress
+      labels:
+        app: wordpress
+    spec:
+      selector:
+        app: wordpress
+      template:
+        metadata:
+          labels:
+            app: wordpress
+        spec:
+          containers:
+            - name: wordpress
+              image: wordpress:4.6.1-apache
+              ports:
+                - containerPort: 80
+status:
+  cluster:
+    namespace: cool
+    name: theperfectkubernetescluster
+  conditions:
+  - lastTransitionTime: 2018-10-02T12:25:39Z
+    lastUpdateTime: 2018-10-02T12:25:39Z
+    message: Successfully submitted cool/supercoolwork
+    status: "True"
+  remote:
+  - apiVersion: extensions/v1beta1
+    kind: Deployment
+    metadata:
+      name: wordpress
+      labels:
+        app: wordpress
+    status:
+      replicas: 2
+      availableReplicas: 2
+      unavailableReplicas: 2
+      observedGeneration: 3
+      conditions:
+      - lastTransitionTime: 2016-10-04T12:25:39Z
+        lastUpdateTime: 2016-10-04T12:25:39Z
+        message: Replica set "nginx-deployment-4262182780" is progressing.
+        reason: ReplicaSetUpdated
+        status: "True"
+        type: Progressing
+```
+
+The monolithic workload design is functionally close to that proposed by this
+document, but has a few major drawbacks:
+
+* It's less composable. Because resource templates are modelled as an array
+  within `KubernetesApplication` a particularly complex application would be
+  _huge_. In the Gitlab example editing a single YAML document containing ~4,800
+  lines could be unwieldy, as would be running
+  `kubectl describe kubernetesapplication gitlab`.
```

Comment:
@ichekrygin Proposed a compromise between the design proposed by this document and monolithic workloads as described in this section that had not occurred to me. His proposal is a little closer to today's pod controllers like `Deployment` in that all templates are written inline in the `KubernetesApplication` `.spec`, but each template results in the creation of a distinct `KubernetesApplicationComponent` resource which can track the status and ownership of the remote object.

For example:
```yaml
---
apiVersion: workload.crossplane.io/v1alpha1
kind: KubernetesApplication
metadata:
  name: demo
spec:
  clusterSelector:
    matchLabels:
      provider: gcp
  componentSelector:
    matchLabels:
      app: demo
  # Each of these templates will be used to create a
  # KubernetesApplicationComponent owned by this KubernetesApplication.
  componentTemplates:
    # I've omitted name here because we could probably generate a name like
    # nameofapplication-namespaceoftemplatedresource-nameoftemplatedresource
  - metadata:
      labels:
        app: demo
    spec:
      secrets:
      - some-connection-secret
      template:
        apiVersion: extensions/v1beta1
        kind: Deployment
        metadata:
          name: wordpress
          labels:
            app: wordpress
        spec:
          selector:
            app: wordpress
          template:
            metadata:
              labels:
                app: wordpress
            spec:
              containers:
                - name: wordpress
                  image: wordpress:4.6.1-apache
                  ports:
                    - containerPort: 80
  - metadata:
      labels:
        app: demo
    spec:
      template:
        apiVersion: v1
        kind: Service
    metadata:
      labels:
        app: wordpress
    spec:
      ports:
        - port: 80
      selector:
        app: wordpress
      type: LoadBalancer
status:
  cluster:
    namespace: cool
    name: theperfectkubernetescluster
  state: Submitted
  desiredComponents: 2
  submittedComponents: 2
  conditions:
  - lastTransitionTime: 2018-10-02T12:25:39Z
    lastUpdateTime: 2018-10-02T12:25:39Z
    message: Successfully submitted KubernetesApplicationComponent default/demo
    status: "True"
```

This compromise eliminates all of the drawbacks mentioned under the Monolithic Workloads heading except for my concerns around huge YAML documents and API objects. The `KubernetesApplication` creates distinct `KubernetesApplicationComponents` from its templates, so we still use those components to model remote ownership and status per this design. It also removes some asymmetry; in my proposed design components and applications are created separately, but the components that match an application's selector are deleted when the application is deleted. So creating an application does not create components, but deleting an application does delete components.

My understanding is that the main motivation behind this compromise is that it reduces the likelihood of label selector conflicts. The concern being that if components are always submitted separately from applications, which subsequently adopt the components that match their label selectors, there's more room for human error. If I created an application selecting label `cool=very` then the next day @ichekrygin tries to create his own different application also selecting label `cool=very` there will be a race between the two applications to adopt @ichekrygin's new components. I've experienced similar issues with `Service` in the wild, in which one person adds a `Service` whose label selectors accidentally match `Pods` that they did not intend to.

I can see @ichekrygin's point here, but I still worry about how huge our `KubernetesApplication` resources are likely to become if they must template complex workloads inline. I wonder if @jbw976 or @bassam can help sway us one way or another.

Reply:
The main motivation is to make Application to be an authoritative source for components definition by including component templates inside the Application spec, similarly to K8S Deployment which includes full specification of the Pod object the Deployment will create. 

In addition to being authoritative in the spec definition, the Application will also become authoritative in terms of Life cycle management (create, update, delete) for the Components. Every modification goes through the Application object, and then cascaded to the components. Again, the very similar analogy to the Deployment - Pod, where user interfaced with Deployment object in CRUD operations and not at the ReplicaSet or Pod level.

Label/Selector mechanism provides a rather weak affinity. Hence, every pod in addition to having labels defined in the spec which match deployment selector has an additional label corresponding to a ReplicaSet hash. This label prevents ReplicaSet to Pod interpolation during update events, when Deployment can have multiple ReplicaSets in operation.

Reply:
To address the size issue. It is very valid and vary good concern about keeping Application.yaml file in manageable size. However, I don't think it should be used in the design making process. Just like Deployment design is un-influenced by number of containers in the pod template, similarly, Application design should not be influenced by the number of the Components.

I totally agree, from the usability perspective it is very important to provide excellend developer experience, but I don't think it should be tailored after exteme cases like GitLab installation. 

Even I am am wrong in "extereme cases" part, and most of the Applications will be at the same complexity level as GitLab, I would suggest that providing `client cli tools` or `kubectl` to manage this is arguable a better way. He is a quick example of Application directory structure for a complex application:
```
ComplexApp/
├── application.yaml
├── component-group-A
│   ├── deployment.yaml
│   └── service.yaml
├── component-group-B
│   ├── configmap.yaml
│   └── job.yaml
├── component-group-C
│   ├── ingress.yaml
│   └── rbac.yaml
├── deployment-blue.yaml
└── service-gree.yaml
```
where all components represented as individual files in either flat or dir-structure layout.
It should be very concieable to come up with simple `cli` command that renders this entire `ComplexApplication` into a single `yaml` file:
`coolctl render -f ComplexApp/ > complex-app.yaml` 
or 
`coolctl render -f ComplexApp/ | kubectl apply  -f -`
or with `kubeclt plugin`
`kubeclt create -f ComplexApp/`, 
etc.

Reply:
> The main motivation is to make Application to be an authoritative source for components definition by including component templates inside the Application spec, similarly to K8S Deployment which includes full specification of the Pod object the Deployment will create.

This sounds more like the solution than the motivation. This is not unimportant detail, but it's additionally useful to capture _why_ applications being authoritative for the entire lifecycle of their components is desirable. Concrete examples like the race between two applications selecting the same labels that you pointed out help a lot here.

So far the pros and cons we've captured for what I'm calling the "compromise design" (as opposed to the "proposed design") are:

Pros:
* Component adoption becomes an edge case. Most components are created with their controller reference set, so there's no risk that someone will accidentally reuse an existing application's label, resulting in a component being controlled by the wrong application.
* There's more symmetry and less surprise. In the proposed design it's surprising that deleting an application deletes its matching components, given that creating an application does not create components.
* It works a little bit more like a Deployment, which is a pattern people are used to. I'm not sure how much we really care about this given that we're doing a different enough thing that we're always going to surprise someone who expects this to work just like how a Deployment works.

Cons:
* It's likely that the KubernetesApplication API resources (and thus YAML documents) will become huge and unwieldy to operate on. You pointed out that this is also possible for Deployment - a Deployment could template a Pod with 20 containers. This is an uncommon case for Deployments, but my gut feeling is that it will not be for KubernetesApplications.

I'm warming to your compromise proposal, but I'm still worried about how awkward it will be to operate on a complicated KubernetesApplication.

Reply:
If we consider the operations/lifecycle angle:
- Compromise: application, the user creates, updates and deletes an application (full stop)
- Proposed: application, the user creates, updates and deletes the application *and* the list of components.

This leads to a realization that the proposed design does not and can not assert on overall application state/status. For me, as a user, it is super important to know if my application is in READY, PENDING or FAILED state.

The key aspect here is determinism in application deployment state:
- Compromise: once an application is submitted (`kubectl apply`), it could exist in the following states:
   - Failed: deployment failed (invalid definition, propagation failure, etc.)
   - Succeeded: deployment succeeded (application and all components have been successfully deployed and propagated)
   - Pending: deployment is in progress (components being propagated)
- Proposed: there is no notion of the sate for the application other than FAILED 

Since in proposed design Application is completely unaware of its components, it cannot assert that it reached either READY or PENDING state. For PENDING state, I am making a distinction between a component that has been submitted and currently in propagation state, vs. a component that has not been submitted yet. The only state Application can assert is the FAILED state, and only in terms of either propagation failure or failed status of the propagated component. 

The application state even more relevant for complex applications like GitLab. For example, say out of 45 components only 30 were submitted (could be multiple reasons for other submissions not going through), and those 30 successfully propagated and up-and-running. That still wouldn't mean that application is READY. This simply wouldn't be possible in the compromised design, where by definition the application submission is either accepted or rejected and once accepted, the application can fully assert that ALL components defined in the application have been successfully propagated and installation/deployment is complete.

Reply:
> For me, as a user, it is super important to know if my application is in READY, PENDING or FAILED state.

I'm using _Submitted_ in place of _Ready_, because the latter implies the application is running and doing something useful, which we are unable to determine in either design. 

> Since in proposed design Application is completely unaware of its components.

It's not completely unaware of its components, but the components are fluid. In both the compromise and proposed designs the application controller is able to assert that "at this point in time I am responsible for N components and M of them have been submitted".

The only difference I can think of is in the case in which a component is malformed - i.e. unable to be submitted to the API server at all. Assume an application consists of three components, one of which is malformed. In the compromise design the entire application would be rejected, presenting the caller with an error. In the proposed design the application and two components would be accepted, resulting in an application that was "submitted" with two desired components and two submitted components. The user would still receive an error informing them about the malformed component when they tried to apply it.

Reply:
> It's not completely unaware of its components, but the components are fluid. In both the compromise and proposed designs the application controller is able to assert that "at this point in time I am responsible for N components and M of them have been submitted".

Let me rephrase, it (proposed design) is unaware of the components that comprise this application. It is only aware of the components that have been already submitted. The current design the application doesn't have any knowledge of how many components there should be in the application, nor what the content of those components. In the proposed design, the application precisely knows how many components there are and what those are. 

Take GitLab example (with hypothetical 45 components, 30 successfully submitted, and 15 MIA) : 
- Proposed design: I am responsible for 30 and 30 has been successfully processed! What does this information convey about the application status? 
- Compromised design: I am responsible for 45 and only 30 has been successfully propagated and remaining 15 are still pending, overall status: PENDING.

> In the compromise design the entire application would be rejected, presenting the caller with an error. In the proposed design the application and two components would be accepted, resulting in an application that was "submitted" with two desired components and two submitted components. The user would still receive an error informing them about the malformed component when they tried to apply it.

Assuming the 3rd component was submitted altogether. What if it wasn't? But let's go with the assumption that it was and errored out.  Where does this leave the application? It is still deployed, it still says "2 out of 2 successfully propagated", but what does this mean? It surely shouldn't mean that the application is READY, but how can I tell that? Unless I saw/catch/capture the error during submission, I wouldn't be able to tell what is going on with the application. Moreover, in order to troubleshoot this, I must figure out which one out of 3 components was not submitted or failed and re-submit it. Now, make it 30 components (or 45 for GitLab sake).

Reply:
I'm exploring the controller implementation for the compromise design. One challenge is determining how to name the templated components when they are created.

ReplicaSets handle this by giving pods random names. They can do this because they only care whether N replicas of an identical pod exist - they don't need to address any one pod specifically. They (I assume) handle scale down by picking a matching pod and deleting it and scale up by adding a new pod. I don't think they update previously created pods when their template is updated.

In our case an application is a big bundle of heterogenous components. On each reconcile loop the application controller must determine which of its templated components must be created, which must be updated, and which must be deleted. We could do this by requiring that each component template specify a name in its metadata. On each sync we either create or update all of our named templates, then find any components that we control but whose name does not exist in our templates and delete them. If we did this we'd need to add a validation webhook to ensure multiple component templates could not be specified with the same name, or we'd end up with one application racing itself trying to create/update the same component twice.

Reply:
> In our case an application is a big bundle of heterogenous components. 

I think it is a list of homogenous components, with each component representing heterogenous payload. It is true, we need to be able to differentiate between those. 

>  If we did this we'd need to add a validation webhook to ensure multiple component templates could not be specified with the same name, or we'd end up with one application racing itself trying to create/update the same component twice.

I :100: agree! Crossplane, in general, is in desperate need for validation/mutation webhooks in a much broader context.

Reply:
> We could do this by requiring that each component template specify a name in its metadata

Or we could simply have them "named",  similar to containers in the pod, each container has a unique name, i.e. it doesn't have to be a full blown `ObjectMetadata` at each component level.

Reply:
Here is a mock example:

Reply:
> Or we could simply have them "named", similar to containers in the pod, each container has a unique name, i.e. it doesn't have to be a full blown ObjectMetadata at each component level.

I'm 👎 on this. It makes sense for containers because they don't exist as first class API resources in Kubernetes. Given we're creating `KubernetesApplicationComponent` resources under the hood I think it makes sense to allow setting metadata for those components. I don't think omitting it gets us much.

Reply:
I think the analogy here is with POD. 

> I don't think omitting it gets us much.

Inversely I think including it gives us same or even less. 

```yaml
...
  componentTemplates:
    - meta: 
        name: foo
        namespace: ? (same as application)
        labels: ? (same as application selector or auto generated)
      spec:
        secrets:
          - some-connection-secret
        template:
          apiVersion: extensions/v1beta1
          kind: Deployment
          metadata:
            name: wordpress
            labels:
              app: wordpress
          spec:
            selector:
              app: wordpress
            template:
              metadata:
                labels:
                  app: wordpress
              spec:
                containers:
                  - name: wordpress
                    image: wordpress:4.6.1-apache
                    ports:
                      - containerPort: 80
    - meta: 
        name: bar
        namespace: ? (same as application)
        labels: ? (same as application selector or auto generated)
      spec:
        template:
          apiVersion: v1
          kind: Service
          spec:
            ports:
              - port: 80
            selector:
              app: wordpress
            type: LoadBalancer
...
```

I am OK either way:
![image](https://user-images.githubusercontent.com/324803/56327292-a1a75a00-612e-11e9-94bb-6baebad3ead8.png)

Reply:
I'm still personally unconvinced that this "compromise" pattern is a clear improvement over the  "loosely coupled" pattern I initially proposed. That said, each pattern has its strengths and weaknesses compared to the other. We're still in alpha, and it's not too difficult to switch from one pattern to the other. Given most maintainers lean toward the compromise pattern I've updated the design document to reflect said pattern and added the original "loosely coupled" pattern to the Alternatives Considered section.